### ISSUE TITLE
do keyword followed by assignment of function definition loses its default value

### ISSUE THREAD

**Issue by ippeiukai** [jashkenas/coffeescript#3776]
``` CoffeeScript
alert do fn = (hoge = 10) -> hoge # 10 as expected
alert fn() # !!! undefined, not 10
```
Very edge case, but I suspect this behaviour is not intended.

**Comment by michaelficarra** [comment]
That's not a default value. `do` has a special form when given a function (or an assignment to a function) directly that passes the "default value" in as an argument. This is basically let aliasing.

**Comment by epidemian** [comment]
This seems like a bug to me (a very interesting one, BTW!).

I personally would expect this:

```coffeescript
console.log do fn = (hoge = 10) -> hoge
```

to be equivalent to:

```coffeescript
console.log do (fn = (hoge = 10) -> hoge)
```

But, alas, it isn't:

```coffeescript
console.log do fn = (hoge = 10) -> hoge # => 10
console.log fn() # => undefined
```

```coffeescript
console.log do (fn = (hoge = 10) -> hoge) # => 10
console.log fn() # => 10
```

Also, given that:

```coffeescript
fn = (hoge = 10) -> hoge
do fn # => 10
fn() => 10
```

works the way it does, and that an assignment is an expression, and that `do` can take any expression, wouldn't it be expected that inlining the assignment inside the `do` should have the same behaviour?

Also, this contradicts [what was said](https://github.com/jashkenas/coffeescript/issues/3772#issuecomment-68877507) in #3772:

> `do` is just function application with zero arguments.

That doesn't seem to be the case

**Comment by ippeiukai** [comment]
@epidemian, Thank you for clearing it up.
@michaelficarra, Can you reopen this issue?

No one would expect results `2b` and `2c`. But I agree with @epidemian that it would be more consistent if `2a` too behaved like `3a`.
Other cases are intuitive though, and like I said this is an edge case. However, like #3772, shouldn't `do` keyword be documented more clearly? It seems to me even a collaborator can get confused about it.

``` CoffeeScript
foo = 0
bar = 1

# ====
# case 0

# 0a: [0, 1]
do (foo, baz = bar) -> 
  console.log [foo, baz]
  undefined

# ====
# case 1

fn1 = (foo, baz = bar) -> 
  console.log [foo, baz]
  undefined

# result 1a: [undefined, 1]
do fn1

# result 1b: [0, 1]
fn1(foo)

# ====
# case 2

# result 2a: [0, 1]
do fn2 = (foo, baz = bar) -> 
  console.log [foo, baz]
  undefined

# result 2b: [undefined, undefined]
do fn2

# result 2c: [0, undefined]
fn2(foo)

# ====
# case 3

# result 3a: [undefined, 1]
do (fn3 = (foo, baz = bar) -> 
  console.log [foo, baz]
  undefined
)

# result 3b: [undefined, 1]
do fn3 

# result 3c: [0, 1]
fn3(foo)
```

**Comment by michaelficarra** [comment]
`do ((a = b) -> a)` not being the same as `do (a = b) -> a` is indeed a bug, but I believe we already have an open issue for that.

**Comment by ippeiukai** [comment]
> do ((a = b) -> a) not being the same as do (a = b) -> a is indeed a bug

I'm now confused. 

Only mention in the documentation I found about `do` keyword is following:
> CoffeeScript provides the do keyword, which immediately invokes a passed function, forwarding any arguments.

This is way too ambiguous.

To start with, I figured there are two ways in which `do` keyword works:
 1. `do` immediately followed by a function definition (`->` or `=>`, optionally following parameters in `()`) executes the function with arguments forwarding.
 2. Otherwise, `do` is a unary operator that executes the value as a function with no argument.

I was assuming `do ((a = b) -> a)` and `do (fn = (a = b) -> a)` should rightly be *2.*, not *1.*. It already works that way too. This issue is about `do fn = (a = b) -> a` not behaving as neither *1.* nor *2.*.

But now you are telling me `do ((a = b) -> a)` should be *1.* instead.
So, what exactly **should** get *1.* and what should *2.* then?
